docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst

```diff
@@ -0,0 +1,356 @@
+
+The data flow analysis is performed using the predicate ``hasFlow(DataFlow::Node source, DataFlow::Node sink)``:
+
+.. code-block:: ql
+
+   from MyDataFlowConfiguation dataflow, DataFlow::Node source, DataFlow::Node sink
+   where dataflow.hasFlow(source, sink)
+   select source, "Dataflow to $@.", sink, sink.toString()
+
+Using global taint tracking
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Global taint tracking is to global data flow what local taint tracking is to local data flow. That is, global taint tracking extends global data flow with additional non-value-preserving steps. The global taint tracking library is used by extending the class ``TaintTracking::Configuration``:
+
+.. code-block:: ql
+
+   import python
+
+   class MyTaintTrackingConfiguration extends TaintTracking::Configuration {
+     MyTaintTrackingConfiguration() { this = "..." }
+
+     override predicate isSource(DataFlow::Node source) {
+       ...
+     }
+
+     override predicate isSink(DataFlow::Node sink) {
+       ...
+     }
+   }
+
+These predicates are defined in the configuration:
+
+-  ``isSource`` - defines where taint may flow from.
+-  ``isSink`` - defines where taint may flow to.
+-  ``isSanitizer`` - optionally, restricts the taint flow.
+-  ``isAdditionalTaintStep`` - optionally, adds additional taint steps.
+
+Similar to global data flow, the characteristic predicate (``MyTaintTrackingConfiguration()``) defines the unique name of the configuration and the taint analysis is performed using the predicate ``hasFlow(DataFlow::Node source, DataFlow::Node sink)``.
+
+Flow sources
+~~~~~~~~~~~~
+
+The data flow library contains some predefined flow sources. The class ``RemoteFlowSource`` (defined in module ``semmle.python.dataflow.new.RemoteFlowSources``) represents data flow from remote network inputs. This is useful for finding security problems in networked services.
+
+For global flow, it is also useful to restrict sources to instances of ``LocalSourceNode``. The predefined sources generally do that.
+
+Example
+~~~~~~~
+
+This query shows a data flow configuration that uses all network input as data sources:
+
+.. code-block:: ql
+
+   import python
+   import semmle.python.dataflow.new.RemoteFlowSources
+
+   class MyDataFlowConfiguration extends DataFlow::Configuration {
+     MyDataFlowConfiguration() {
+       this = "..."
+     }
+
+     override predicate isSource(DataFlow::Node source) {
+       source instanceof RemoteFlowSource
+     }
+
+     ...
+   }
+
+Class hierarchy
+~~~~~~~~~~~~~~~
+
+-  ``DataFlow::Configuration`` - base class for custom global data flow analysis.
+-  ``DataFlow::Node`` - an element behaving as a data flow node.
+
+   -  ``DataFlow::ExprNode`` - an expression behaving as a data flow node.
+   -  ``DataFlow::ParameterNode`` - a parameter data flow node representing the value of a parameter at function entry.
```

Comment:
```suggestion
   -  ``DataFlow::ExprNode`` - an expression behaving as a data flow node.
   -  ``DataFlow::CfgNode`` - a control-flow node behaving as a data flow node.
   -  ``DataFlow::CallCfgNode`` - a control-flow node for a function or method call behaving as a data flow node.
   -  ``DataFlow::ParameterNode`` - a parameter data flow node representing the value of a parameter at function entry.
```

Reply:
Good suggestion, I have incorporated `CfgNode` and `CallCfgNode` into the hierarchy.